File: Framework/Php/framework/framework/runtime/smartserialization.php
<?php

/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */

namespace Smartkernel\Framework\Runtime {

    use Smartkernel\Framework\IO\SmartPath;
    use Smartkernel\Framework\IO\SmartDirectory;
    use Smartkernel\Framework\IO\SmartFile;
    use Smartkernel\Framework\Text\SmartEncoding;

/*
     * 序列化
     */

    class SmartSerialization {
        /*
         * 序列化过程
         * 
         * @param object $obj，准备序列化的对象
         * @param string $filePath，文件路径，文件不存在时会自动创建
         * @param $encoding，编码方式
         * @param $filePathEncoding，文件路径编码
         * 
         */

        public static function serialize($obj, $filePath, $encoding = null, $filePathEncoding = SmartEncoding::DefaultSafePathValue) {
            $directoryName = SmartPath::getDirectoryName($filePath);
            if (!SmartDirectory::exists($directoryName)) {
                SmartDirectory::createDirectory($directoryName);
            }
            SmartFile::write($filePath, serialize($obj), false, $encoding, $filePathEncoding);
        }

        /*
         * 反序列化过程
         * 
         * @param string $filePath，读取序列化数据的文件
         * @param $encoding，编码方式
         * @param $filePathEncoding，文件路径编码
         * @return object，返回反序列化的对象
         * 
         */

        public static function deserialize($filePath, $encoding = null, $filePathEncoding = SmartEncoding::DefaultSafePathValue) {

            return unserialize(SmartFile::readAllText($filePath, $encoding, $filePathEncoding));
        }

        /*
         * 克隆对象
         * 
         * @param object $obj，对象
         * @return object，克隆之后的对象
         * 
         */

        public static function cloneObject($obj) {
            return unserialize(serialize($obj));
        }

    }

}

File: Framework/CSharp/Framework/Framework/Runtime/SmartSerialization.cs
/*
 * 项目：Smartkernel.Framework
 * 作者：曹艳白
 * 邮箱：caoyanbai@139.com
 */
using System;
using System.IO;
using System.Runtime.Serialization;
using System.Runtime.Serialization.Formatters.Binary;
using System.Runtime.Serialization.Formatters.Soap;

namespace Smartkernel.Framework.Runtime
{
    /// <summary>
    /// 序列化
    /// </summary>
    public static class SmartSerialization
    {
        /// <summary>
        /// 二进制序列化器
        /// </summary>
        private static readonly BinaryFormatter binaryFormatter = new BinaryFormatter();

        /// <summary>
        /// 克隆对象
        /// </summary>
        /// <typeparam name="T">对象类型</typeparam>
        /// <param name="obj">对象</param>
        /// <returns>克隆之后的对象</returns>
        public static T Clone<T>(object obj) where T : new()
        {
            using (var memoryStream = new MemoryStream())
            {
                binaryFormatter.Serialize(memoryStream, obj);
                memoryStream.Position = 0;
                return (T)binaryFormatter.Deserialize(memoryStream);
            }
        }

        /// <summary>
        /// 序列化过程
        /// </summary>
        /// <param name="obj">准备序列化的对象</param>
        /// <param name="stream">序列化对象写入的流</param>
        /// <param name="serializationFormatType">序列化的类型</param>
        public static void Serialize(Object obj, Stream stream, SmartSerializationFormatType serializationFormatType = SmartSerializationFormatType.Binary)
        {
            IFormatter formatter = null;
            switch (serializationFormatType)
            {
                case SmartSerializationFormatType.Soap:
                    formatter = new SoapFormatter();
                    break;
                case SmartSerializationFormatType.Binary:
                    formatter = new BinaryFormatter();
                    break;
                default:
                    formatter = new BinaryFormatter();
                    break;
            }
            formatter.Serialize(stream, obj);
            stream.Position = 0;
        }

        /// <summary>
        /// 序列化过程
        /// </summary>
        /// <param name="obj">准备序列化的对象</param>
        /// <param name="filePath">序列化写入的文件</param>
        /// <param name="serializationFormatType">序列化的类型</param>
        public static void Serialize(Object obj, string filePath, SmartSerializationFormatType serializationFormatType = SmartSerializationFormatType.Binary)
        {
            var directoryName = Path.GetDirectoryName(filePath);
            if (!Directory.Exists(directoryName))
            {
                Directory.CreateDirectory(directoryName);
            }
            using (Stream stream = new FileStream(filePath, FileMode.OpenOrCreate))
            {
                Serialize(obj, stream, serializationFormatType);
            }
        }

        /// <summary>
        /// 反序列化过程
        /// </summary>
        /// <typeparam name="T">反序列化的目标类型</typeparam>
        /// <param name="stream">序列化对象的流</param>
        /// <param name="serializationFormatType">序列化的类型</param>
        /// <returns>返回反序列化的对象</returns>
        public static T Deserialize<T>(Stream stream, SmartSerializationFormatType serializationFormatType = SmartSerializationFormatType.Binary)
        {
            IFormatter formatter = null;
            switch (serializationFormatType)
            {
                case SmartSerializationFormatType.Soap:
                    formatter = new SoapFormatter();
                    break;
                case SmartSerializationFormatType.Binary:
                    formatter = new BinaryFormatter();
                    break;
                default:
                    formatter = new BinaryFormatter();
                    break;
            }
            return (T)formatter.Deserialize(stream);
        }

        /// <summary>
        /// 反序列化过程
        /// </summary>
        /// <typeparam name="T">反序列化的目标类型</typeparam>
        /// <param name="filePath">读取序列化数据的文件</param>
        /// <param name="serializationFormatType">序列化的类型</param>
        /// <returns>返回反序列化的对象</returns>
        public static T Deserialize<T>(string filePath, SmartSerializationFormatType serializationFormatType = SmartSerializationFormatType.Binary)
        {
            using (Stream stream = new FileStream(filePath, FileMode.Open))
            {
                return Deserialize<T>(stream, serializationFormatType);
            }
        }
    }
}
